src/coreclr/jit/emitriscv64.cpp

```diff
@@ -566,33 +567,48 @@ void emitter::emitIns_R_R(
         code |= reg1 << 7;
         code |= reg2 << 15;
     }
-    else if ((INS_fcvt_w_s <= ins && INS_fmv_x_w >= ins) || (INS_fclass_s == ins || INS_fclass_d == ins) ||
-             (INS_fcvt_w_d == ins || INS_fcvt_wu_d == ins) || (INS_fcvt_l_s == ins || INS_fcvt_lu_s == ins) ||
-             (INS_fmv_x_d == ins))
+    else if (INS_fmv_x_d == ins || INS_fmv_x_w == ins || INS_fclass_s == ins || INS_fclass_d == ins)
     {
-        // TODO-RISCV64-CQ: Check rounding mode
         assert(isGeneralRegisterOrR0(reg1));
         assert(isFloatReg(reg2));
-        code |= (reg1 & 0x1f) << 7;
+        code |= reg1 << 7;
         code |= (reg2 & 0x1f) << 15;
     }
-    else if ((INS_fcvt_s_w <= ins && INS_fmv_w_x >= ins) || (INS_fcvt_d_w == ins || INS_fcvt_d_wu == ins) ||
-             (INS_fcvt_s_l == ins || INS_fcvt_s_lu == ins) || (INS_fmv_d_x == ins) ||
-             (INS_fcvt_d_l == ins || INS_fcvt_d_lu == ins))
-
+    else if (INS_fcvt_w_s == ins || INS_fcvt_wu_s == ins || INS_fcvt_w_d == ins || INS_fcvt_wu_d == ins ||
+             INS_fcvt_l_s == ins || INS_fcvt_lu_s == ins || INS_fcvt_l_d == ins || INS_fcvt_lu_d == ins)
     {
+        assert(isGeneralRegisterOrR0(reg1));
+        assert(isFloatReg(reg2));
+        code |= reg1 << 7;
+        code |= (reg2 & 0x1f) << 15;
         // TODO-RISCV64-CQ: Check rounding mode
+        code |= 0x1 << 12;
```

Comment:
Is it correct rounding mode for .net runtime? Please help me!

Reply:
Ping @tannergooding, can you help with the right rounding mode to use for these floating point conversions?

Reply:
@tannergooding  Could you please review?
I checked that arm64 and loongarch64 used `Towards Zero` only for FloatRegToGeneralReg. So I set `0x1` for rounding mode in this code only. A manual which I referenced is https://five-embeddev.com/riscv-isa-manual/latest/f.html
Thank you.

Reply:
Sorry, I missed this.

.NET does not support IEEE 754 exceptions and does not support it being changed. That is, sNaN should not trigger exceptions, inexact results (including overflow/underflow) should not trigger exceptions, invalid operations (such as `0 * Infinity` or `division by 0`) should not trigger exceptions.

.NET only supports `Round to Nearest, ties to Even` for standard operations and does not support it being changed.

For conversion to integral types, it uses "truncation" sometimes referred to as "Round towards Zero".

For certain `Math` functions it does allow alternative conversion to integral values:
* `Floor` - Toward Negative Infinity
* `Ceiling` - Toward Positive Infinity
* `Truncate` - Towards Zero
* `Round` - To nearest, ties to even
* `Round + MidpointRounding` - Depends on the selected `MidpointRounding` value

---------------------

For RISC-V, this should be handled by initial population of `11.2 Floating-Point Control and Status Register` to be `0b000_00000`. This will set the rounding mode to `Round to Nearest, ties to Even`. RISC-V already does not support triggering IEEE 754 exceptions and simply provides a way to check the flags using the lowest 5 bits so software can handle things instead.

For conversion to integers, `11.7 Single-Precision Floating-Point Conversion and Move Instructions` covers the general needs and will round according to their `rm` field. The `rm` field should default in this case to `0b001` which will select `Round towards Zero`. As an optimization you could also recognize `CAST(ROUND(x))` and change the RM field to be correct.

I do not, at a glance, see any instructions which `round` but preserve the value as a floating-point value.

Reply:
@tannergooding As I understand, you mean that 1) normal rounding is handled by status register (its initial value is `Round to Nearest, ties to Even`) in RISC-V and 2) for conversion to integers rounding uses `Round towards zero`. 3) Both .NET and RISC-V do not support IEEE-754 exceptions. So I think I correctly set rounding mode to the instructions in this PR, right?

And I have a question, should I set initial value to status register in the runtime code? If then, could you explain where is the correct location in the runtime source code? Actually, I tried to find the code location where initial value of status register is set in ARM64 and LOONGARCH. I couldn't find, 

Thank you. If I misunderstand your explanation, please let me know.

Reply:
We do not set floating-point mode on startup. We assume platform default setting. Changing the global floating-point mode is a problem for interop. It is not feasible to set/reset it at each interop boundary.

We reset the floating mode to the default setting on Windows only when we are initializing a new thread: https://github.com/dotnet/runtime/blob/2c9e48366b20d0d4b398e9c1d114c7cc1a4e5f83/src/coreclr/vm/threads.cpp#L1751 . This is left-over from .NET Framework for bug-for-bug compatibility. I think it would be ok to remove it.

Reply:
@tannergooding @jkotas Thank you!

Reply:
@clamp03 Can you remove the TODOs related to this if this is resolved now?

Reply:
I already removed in https://github.com/dotnet/runtime/pull/84748/commits/1946f91171f95982b8f28ff28e9a07e3a1cb1e70 commit. Hm... I don't know why TODOs in comments are not removed.

Reply:
Ah, sorry, I think these reviews are just attached to outdated code then.